libs/db/src/prisma/schema.prisma

```diff
@@ -374,6 +379,174 @@ model MapReviewComment {
+/// `gamemode`, `trackType`, `trackNum`, `style`). By using that combination in
+/// our PK, we ensure those entries are unique at a DB level.
+///
+/// We use B-Tree two indexes
+///   - (userID, gamemode, style, mapID, trackType, trackNum): For fetching runs by user for e.g. profile page.
+///     Also is a UNIQUE constraint.
+///   - (mapID, gamemode, trackType, trackNum, style, time (asc), createdAt (asc)): Indexes in leaderboard order.
+///     The idea is we get every distinct leaderboard stacked on top of each other, then ordered by time within the chunk,
+///     where ties are simply handled by whatever was submitted first. Note that we store any actual "rank" values - incrementing
+///     ranks for all slower runs on a leaderboard is an incredibly slow operation. (TODO: This hasn't been implemented yet).
+///     Instead, we index this way to allow us to very efficiently compute ranks per query using window functions. 
+model LeaderboardRun {
+  leaderboard Leaderboard @relation(fields: [mapID, gamemode, trackType, trackNum, style], references: [mapID, gamemode, trackType, trackNum, style], onDelete: Cascade)
+
+  user   User @relation(fields: [userID], references: [id], onDelete: Restrict)
+  userID Int
+
+  mmap  MMap @relation(fields: [mapID], references: [id], onDelete: Cascade)
+  mapID Int
+
+  gamemode  Int @db.SmallInt
+  trackType Int @db.SmallInt
+  trackNum  Int @db.SmallInt
+  style     Int @db.SmallInt
+
+  /// We don't store ticks in anticipation of future sub-tick system. If ticks 
+  /// are required, they can be calculated with `time * tickrate of gamemode`.
+  time Float @db.Real
+
+  /// We store this for LeaderboardRun and *not* PastRun for space efficiency.
+  stats Json /// run-stats.model.ts
+
+  // validated  Boolean // TODO: We'll be using this in future and blocking a bunch of stuff from DTO if false
+  replayHash String? @db.Char(40) /// SHA1
+
+  /// Flags are all the styles are run qualifies for. A styled run may qualify
+  /// for multiple leaderboards, and an separate entry is made for each. 
+  /// For example, a sideways surf run (for some track) would be added to the
+  /// sideways leaderboard and unstyled leaderboard. This field is used to "tag"
+  /// runs one leaderboard as also qualifying for another, e.g. the sideways 
+  /// run would get a "Sideways" tag on the unstyled leaderboard.
+  // This is an array because (a) it allows easier queries for specific flag(s)
+  // in Prisma (writing the paginated PastRuns get query raw would be really
+  // rough, though we could if absolutely needed), (b) we could conceivable
+  // exceed 64 flags at some point. This is of course a very large table where
+  // we care about size a great deal, though we expect the *majority* of runs
+  // If we need to cut back on the size of this table in the future, we could
+  // submitted not to have any flags, so it's probably pretty okay.
```

Comment:
Looks like these comment lines got swapped

Reply:
Or rather I just started deliriously rambling - update comment should make more sense.

Reply:
It makes perfect sense if they were swapped!

Reply:
I'm confused, can you write exactly what you mean?

To expand on the comment a bit, Prisma schemas don't supposed nullable arrays for some reason (it's a much requested), so I'm not sure what Prisma ends up entering into the table when using an array. Postgres will use a bit more data for empty arrays over NULL (https://dba.stackexchange.com/questions/223518/is-there-a-disk-space-impact-when-choosing-null-vs-empty-array-for-the-default) so there may be a small size benefit to a bitfield, even for runs with no flags (likely quite a major of runs). If we added pguint (considering it) this could be some array of 1-byte uints which is probably very space-efficient (Postgres arrays use `varlena` under the hood which is IIRC on disk is essentially just an integer for the number of items followed by the items themselves), maybe in some cases runs with a single 1-byte flag plus extra array stuff ends up less than the 8 bytes for the bitfield.  Plus we could *maybe* have >64 styles (most we could pack in a single int) but unlikely to ever be >256.

Sorry dunno how much sense that makes, half asleep so more delirious rambling I suppose

Reply:
Don't worry about it, it's not a big deal and I'm not seeing what I saw before anyway (hopefully I didn't imagine it).

As for the best way to store tags, it seems like all the options are similar size-wise in the grand scheme of things, so I would just go with whatever is easiest to work with. I doubt we will ever have very many styles anyway, especially since there are downsides to having tons of small leaderboards.

Also, I think all runs should be tagged with their style, including "normal" runs. Otherwise it might be tricky to filter for those runs, especially if other non-style tags are present. And eventually, "normal" will be a sub-style of segmented and TAS, so we can't necessarily omit the tag due to it being the "top level" style.

Reply:
Mhmm okay I suppose you're right - maybe programmatically to omit the normal style and infer whether or not it exists based on the Run has tags incompatible with it (eg if it's not TAS segmented but, can't we infer it's normal?). But def sounds annoying do account for across multiple codebases and it's a very small space optimisation that we could always make further down the line if we got really desperate about space.